### PR TITLE
Implement lexer

### PR DESCRIPTION
**Pull request by PascalHonegger**

### CODE REVIEW

File: common/src/main/kotlin/common/Diagnostics.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package common
+
+enum class DiagnosticSeverity {
+    Error, Warning, Info
```
Comment (mjossdev) [inline common/src/main/kotlin/common/Diagnostics.kt:4]: Enum values in Java and Kotlin are typically CAPS_LOCK

https://kotlinlang.org/docs/enum-classes.html

File: common/src/main/kotlin/common/Diagnostics.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+data class DiagnosticMessage(val message: String, val location: Location, val severity: DiagnosticSeverity) {
+    override fun toString(): String {
+        return "$severity: $message at $location"
+    }
+}
+
+class DiagnosticConsole {
+    private val mutableMessages: MutableList<DiagnosticMessage> = ArrayList()
+    val messages: List<DiagnosticMessage> = mutableMessages
```
Comment (mjossdev) [inline common/src/main/kotlin/common/Diagnostics.kt:15]: Could be a getter
```suggestion
    val messages: List<DiagnosticMessage> get() = mutableMessages
```

File: lexer/src/main/kotlin/lexer/Lexer.kt [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            }
+            val hasWholeDigits = readDigits()
+            if (!hasWholeDigits) {
+                diagnostics.reportError("Number is missing whole part (0.5 is valid, .5 is not)", locationFromStart(start))
+                append('0')
+            }
+
+            if (current == '.') {
+                reader.mark(1)
+                assert(reader.markSupported())
```
Comment (mjossdev) [inline lexer/src/main/kotlin/lexer/Lexer.kt:99]: mark would throw an IOException anyways if it's not supported
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/Reader.html#mark(int)

File: common/src/main/kotlin/common/Diagnostics.kt (hunk 1 above)
Comment (PascalHonegger) [reply common/src/main/kotlin/common/Diagnostics.kt:4]: According to the latest [Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html), both are fine:

> For enum constants, it's OK to use either uppercase underscore-separated names ([screaming snake case](https://en.wikipedia.org/wiki/Snake_case)) (enum class Color { RED, GREEN }) or upper camel case names, depending on the usage.

I personally prefer `PascalCase`, hence why it's used here

File: common/src/main/kotlin/common/Diagnostics.kt (hunk 2 above)
Comment (PascalHonegger) [reply common/src/main/kotlin/common/Diagnostics.kt:15]: Changed to getter

File: lexer/src/main/kotlin/lexer/Lexer.kt (hunk 3 above)
Comment (PascalHonegger) [reply lexer/src/main/kotlin/lexer/Lexer.kt:99]: Removed the `assert`

**Review by mjossdev — APPROVED** [review]

**Pull request merged**